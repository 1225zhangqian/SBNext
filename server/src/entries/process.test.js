import { readFileSync, statSync } from 'fs';
import fm from 'frontmatter';
import processEntries from './process';

jest.mock('fs');
jest.mock('frontmatter');

describe('processEntries ', () => {
  beforeAll(() => {
    const constantDate = new Date('2017');

    global.Date = class extends Date {
      constructor() {
        super();
        return constantDate;
      }
    };
  });

  test('exports processEntries fn as default', () => {
    expect(processEntries).toBeDefined();
  });

  test('retrieves an array of docs with default values', async () => {
    const expectedName = 'test';
    const expectedEntry = `docs/${expectedName}.md`;
    const expectedContent = 'text';
    const expectedDate = new Date();
    const expectedFileContent = `
        ---
        ---
        ${expectedContent}
        `;

    readFileSync.mockReturnValueOnce(expectedFileContent);

    fm.mockReturnValueOnce({
      data: {},
      content: expectedContent,
    });

    statSync.mockReturnValueOnce({
      birthtime: expectedDate,
    });

    const actual = await processEntries([expectedEntry], 'docs');

    expect(actual).toMatchSnapshot();
  });
});

describe('frontmatter: permalink', () => {
  test('default permalink', async () => {
    const expectedCategory = 'category';
    const expectedName = 'test';
    const expectedEntry = `docs/${expectedName}.md`;
    const expectedDate = new Date();

    fm.mockReturnValueOnce({
      data: {
        category: expectedCategory,
      },
    });

    statSync.mockReturnValueOnce({
      birthtime: expectedDate,
    });

    const actual = await processEntries([expectedEntry], 'docs');

    expect(actual).toMatchSnapshot();
  });

  test('permalink /:category/:name.html', async () => {
    const permalink = '/:category/:name.html';
    const expectedCategory = 'category';
    const expectedName = 'test';
    const expectedEntry = `docs/${expectedName}.md`;
    const expectedDate = new Date();

    fm.mockReturnValueOnce({
      data: {
        category: expectedCategory,
        permalink,
      },
    });

    statSync.mockReturnValueOnce({
      birthtime: expectedDate,
    });

    const actual = await processEntries([expectedEntry], 'docs');

    expect(actual).toMatchSnapshot();
  });

  test('permalink /:date/:name.html', async () => {
    const permalink = '/:date/:name.html';
    const expectedCategory = 'category';
    const expectedName = 'test';
    const expectedEntry = `docs/${expectedName}.md`;
    const dateStr = '2017-05-01';
    const expectedDate = new Date(dateStr);

    fm.mockReturnValueOnce({
      data: {
        category: expectedCategory,
        permalink,
        date: expectedDate,
      },
    });

    statSync.mockReturnValueOnce({
      birthtime: expectedDate,
    });

    const actual = await processEntries([expectedEntry], 'docs');

    expect(actual).toMatchSnapshot();
  });
});
