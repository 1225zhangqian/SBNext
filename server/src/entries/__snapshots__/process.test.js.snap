// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontmatter: permalink default permalink 1`] = `
Array [
  Object {
    "data": Object {
      "_entry": "docs/test.md",
      "category": "category",
      "date": "2017-01-01T00:00:00.000Z",
      "name": "test",
      "url": "/docs/test/",
    },
  },
]
`;

exports[`frontmatter: permalink permalink /:category/:name.html 1`] = `
Array [
  Object {
    "data": Object {
      "_entry": "docs/test.md",
      "category": "category",
      "date": "2017-01-01T00:00:00.000Z",
      "name": "test",
      "permalink": "/:category/:name.html",
      "url": "/docs/test/",
    },
  },
]
`;

exports[`frontmatter: permalink permalink /:date/:name.html 1`] = `
Array [
  Object {
    "data": Object {
      "_entry": "docs/test.md",
      "category": "category",
      "date": "2017-01-01T00:00:00.000Z",
      "name": "test",
      "permalink": "/:date/:name.html",
      "url": "/docs/test/",
    },
  },
]
`;

exports[`processEntries  retrieves an array of docs with default values 1`] = `
Array [
  Object {
    "content": "text",
    "data": Object {
      "_entry": "docs/test.md",
      "date": "2017-01-01T00:00:00.000Z",
      "name": "test",
      "url": "/docs/test/",
    },
  },
]
`;
